### ISSUE TITLE
Enhance the CachingMemoryManager for low mem usecase

### ISSUE THREAD

**Issue by WilliamTambellini** [facebookresearch/flashlight#180]
### Feature Description
As today the CachingMemoryManager is faster than the default AF mem mgr but greedy in memory preventing to train big models on low memory GPUs. This feature proposes to add 2 new params to the CachingMemMgr in order to bypass this limitation. 

#### Use Case
Scenario showing the limit of the current CachingMemMgr, considering a 12GB GPU:
- allocate an array of 10GB
- release the variable/AF array : the 10GB mem buf is not free but just registered to the large free pool ("largeBlocks_") for later recycling
- create a 5GB array : see the CachingMemMgr recycles the previous 10GB buffer by splitting the block in 2 creating a free block of 5GB for the remaining right side of the buffer:
https://github.com/facebookresearch/flashlight/blob/e692a941e902f9e207a1a4ee226c0ad0f8dbdc5e/flashlight/memory/managers/CachingMemoryManager.cpp#L149
- try to create a 6GB array : the available 5GB block is too small and cannot be free because being a 'split', the GPU still has 2GB of unallocated mem, meaning the GPU still has total 7G of unused mem (5+2), but see that the CachingMemMgr OOM. 
This proposal would allow to bypass this limitation. 

[Example usage of the feature - why is it needed?]
It is needed in order to train bigger model without the need for a high mem capacity GPU.
I have found at least 1 way to enhance the CachingMemMgr to be able to handle such scenario by just adding 2 new params to be set accordingly by the user, the default value not changing the current behavior :
size_t recyclingSizeLimit = std::numeric_limit<size_t>::max(); // prevent to recycle buffers, to be set by the user if desired
size_t splitSizeLimit = std::numeric_limit<size_t>::max(); // prevents to split big buffers, to be set by the user if desired
to be used in CachingMemoryManager::alloc(...):
https://github.com/facebookresearch/flashlight/blob/e692a941e902f9e207a1a4ee226c0ad0f8dbdc5e/flashlight/memory/managers/CachingMemoryManager.cpp#L111

1 in order to control the recycling gate : 
https://github.com/facebookresearch/flashlight/blob/e692a941e902f9e207a1a4ee226c0ad0f8dbdc5e/flashlight/memory/managers/CachingMemoryManager.cpp#L130
 if (it != pool.end() && ( isSmallAlloc || (*it)->size_ < recyclingSizeLimit)) {

2 in order to prevent the split of big buffers generating fragmentation:
https://github.com/facebookresearch/flashlight/blob/e692a941e902f9e207a1a4ee226c0ad0f8dbdc5e/flashlight/memory/managers/CachingMemoryManager.cpp#L148
 if ((block->size_ < splitSizeLimit) // dont split big buffers
   && (diff >= (isSmallAlloc ? kMinBlockSize : kSmallSize))) {

By adding these 2 new params and fine setting it accordingly (in my case, recyclingLimit = 400MB and splitLimit=2GB), I ve been able to train some big models on a 12GB GPU where the current CachingMemMgr without these 2 options was needing a 24GB GPU.
Reminder: the default values (max) dont change the current behavior.

Note:
For the cuda backend, a long term enhancement would be to take advantage of nvidia vmm in order to potentially defragment the buffers but it would mean many changes in perhaps both AF and probably FL: 
https://developer.nvidia.com/blog/introducing-low-level-gpu-virtual-memory-management/

Considering the default behavior would not change, would you accept a PR for me to add these 2 params?
Kind
W.

**Comment by jacobkahn** [comment]
@WilliamTambellini — thanks for this writeup. This is a good idea and this approach seems promising. A few comments:
1. It seems that `recyclingSizeLimit` and `splitSizeLimit` would be configurable from outside the memory manager, right? This design makes the most sense to me.
2. These seem like they are sensitive to the allocation pattern and thus the type of model. Can you share some more information about your benchmarks?

Re the CUDA backend and VMM — this has been on my radar for quite some time. You are right; the defragmentation benefits, especially the virtual manager is able to grow or pick better-fitting chunks as needed, might make obsolete some of the code we have. We should talk about it further — we've used CUDA 10.2+ for a little while now.

tl;dr — a PR on these things would be great if you have promising results.

**Comment by WilliamTambellini** [comment]
Tks @jacobkahn 
1 Yes, these 2 variables should better be non-const and easily configurable in order to allow the user to set it from his main/env/config files. I would actually prefer to allow setting it at constructor time with 2 environment variables : for instance: FL_CMM_RECSIZELIMIT and FL_CMM_SPLITSIZELIMIT, in order to play with these params with the default FL build, Would you mind?  Making these params non const opens the door to set these params dynamically at runtime in order for the manager to adapt to the model/mem pressure (for instance by decreasing these limits when OOM hits).

2 For confidentiality reasons, I can not share the details of the models I ve tested with but any model with a large pre softmax layer relative to any 'large' vocabulary size with a high batchsize and/or just a high number of layers should reproduce this issue. Or simply running any training on a small GPUs with only 4G or even 6G of memory would tease the CachingMemMgr too. Wondering if these new options could help that issue : 
https://github.com/facebookresearch/wav2letter/issues/766

On my side, as said, these 2 new parameters do allow me to train my (biggest) models on 12GB GPUs. Could I now prepare a PR?
Kind
W.

**Comment by jacobkahn** [comment]
@WilliamTambellini —
1) Those params should be non-`const`, I agree. I think the approach we've taken previously with setting some of these parameters is with flags that are passed to training scripts, e.g. our old `memstepsize` flag in wav2letter which called `af::setMemStepSize`, but we can also add environment variables to do this, sure. Setting at runtime is preferable in my opinion.
2) No problem re sharing those results, and this makes sense. I'll take a look at your PR.
